alns/select/WeightScheme.py

```diff
@@ -0,0 +1,76 @@
+from abc import abstractmethod
+from typing import List, Optional
+
+import numpy as np
+
+from alns.select.SelectionScheme import SelectionScheme
+
+
+class WeightScheme(SelectionScheme):
+    """
+    Base class from which to implement an operator selection scheme based
+    on weights.
+
+    Parameters
+    ----------
+    scores
+        A list of four non-negative elements, representing the weight
+        updates when the candidate solution results in a new global best
+        (idx 0), is better than the current solution (idx 1), the solution
+        is accepted (idx 2), or rejected (idx 3).
+    num_destroy
+        Number of destroy operators.
+    num_repair
+        Number of repair operators.
+    op_coupling
+        Optional keyword argument. Matrix that indicates coupling between
+        destroy and repair operators. Entry (i, j) is 1 if destroy operator i
+        can be used in conjunction with repair operator j and 0 otherwise.
+    """
```

Comment:
The currnet `WeightScheme` is restrictive in that it assumes there are 4 possible outcomes. Updating is only based on these outcomes and not on objective gain (see comment above).

I think we can remove `WeightScheme` entirely and make it part of `RouletteWheel`. Any other selection scheme based on weights can just inherit directly from `SelectionScheme`.

Reply:
I think so too. I need to think a bit more about how this should work. That probably also depends on what's out there. Maybe we should take a step back, and first find a few more examples in the literature of different selection methods?

I know there's quite a bit ML-based stuff in here too, although I do not have references at the ready. We should probably provide enough context for such ML-based methods to be implementable as well.

I will have a further look tomorrow.